### PR TITLE
ci: add single file workflows for apps and libs

### PR DESCRIPTION
**Pull request by varl**
## Background

This thread on the DHIS2 Slack: https://dhis2.slack.com/archives/CPGUFVC56/p1599658597261100

TL;DR: Reduce our CO2 emissions by avoiding unnecessary CPU work.

## Outline

- ~Do not run workflow on each commit pushed to the repository.~
  Run the workflow on each push to a branch.
- Run the workflow on pull requests, skipping the **publish** job.
- Only **publish** when pushing to authorized branches.
- ~Cache the Yarn cache for subsequent jobs.~
  Cache the node_modules for subsequent jobs.
- **install** job runs first, blocking parallel jobs (**test**, **lint**) until the cache has been created.
- Subsequent jobs must restore the cache~, and `yarn install` from the restored cache.~
- **publish** will run after **install**, **test**, and **lint** have all completed successfully.

## Changes

- ~For apps, we no longer publish _each_ commit to https://github.com/d2-ci, only the ones that satisfy the requirements for the **publish** job, we may need to add additional branches (e.g. `v33`, `patch/2.35.0`, etc.) to the accept list.~

## Result

We can drop-in `dhis2-verify-app.yml` into apps and `dhis2-verify-lib.yml` into libs and they should provide sane defaults.

We can add support for workflows into `d2-style`, so the `d2-style install` commands knows about workflows in this repo: `d2-style install workflow/app`.

## Why not use a custom action (right now)?

Actions cannot be defined using workflow definitions in YAML, they need to be programmatic (JavaScript), which would be the next iteration to explore.

## Demos

- Node (no build): https://github.com/dhis2/cli-style
- Lib (monorepo): https://github.com/dhis2/ui
- App: https://github.com/dhis2/usage-analytics-app

### CODE REVIEW

File: ci/dhis2-verify-lib.yml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+            - name: Install
+              run: yarn install --frozen-lockfile
+
+            - name: Test
+              run: yarn test
+
+    publish:
+        runs-on: ubuntu-latest
+        needs: [install, lint, test]
```
Comment (amcgee) [inline ci/dhis2-verify-lib.yml:114]: Libs also need to be build before publication, I believe?  Looks like the lib workflow doesn't have a build step

File: ci/dhis2-verify-app.yml [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            - master
+            - next
+            - next-major
+            - alpha
+            - beta
+            - '[0-9]+.x'
+            - '[0-9]+.x.x'
+            - '[0-9]+.[0-9]+.x'
+
+    pull_request:
```
Comment (amcgee) [inline ci/dhis2-verify-app.yml:17]: Does this work on to push feature branches to d2-ci when on a pull request branch?  I assume it does...

File: ci/dhis2-verify-app.yml [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                  key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+                  restore-keys: |
+                      ${{ runner.os }}-yarn-
+
+            - name: Install
+              run: yarn install --frozen-lockfile
+
+            - uses: dhis2/deploy-build@master
+              env:
+                  INPUT_BUILD_DIR: build/app
```
Comment (amcgee) [inline ci/dhis2-verify-app.yml:167]: I'm pretty sure this directory won't exist here, since the file system isn't shared between jobs.  In order to transfer the build output artifact from the `build` job to the `publish` job, we'll need to create (and download) a build artifact 

https://docs.github.com/en/free-pro-team@latest/actions/guides/storing-workflow-data-as-artifacts

**Review by amcgee — CHANGES_REQUESTED** [review]
Nice, I really like having these in one workflow even if the files are a little less readable!

I think we could probably create a composite action which does our checkout / install / cache sequence to avoid duplicating that every time, but that's an optimization for another day.

I made a couple suggestions about the sequence of jobs, as written (unless I'm missing something which is possible) I don't think the build output will be available to the publish job in either workflow

**Comment by amcgee** [comment]
> Actions cannot be defined using workflow definitions in YAML, they need to be programmatic (JavaScript), which would be the next iteration to explore.

I *think* we can use a [composite run steps action](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action) for this?  (Haven't tested or looked into it more than noting this doc)

**Comment by amcgee** [comment]
> TL;DR: Reduce our CO2 emissions by avoiding unnecessary CPU work.

I'm ALL about this.  However, a more practical and awesome benefit of this (at least the caching) is that action verifications finish much faster when jobs are run in parallel and caches are used, so we save 10s of seconds to minutes on pull request check runs for each commit 🎉 

> install job runs first, blocking parallel jobs (test, lint) until the cache has been created.

Were you able to test this?  I'm not sure we gain much by saving and restoring the Yarn cache between jobs, as the install job itself still needs to run for each job and now it actually needs to run twice before any job can finish.  Maybe we could cache `node_modules` in the `install` job and drop the `yarn install` repeated for each other job instead?  Alternatively, just having the cache action and `yarn install` for each of the `lint`/`test`/`build`/`publish` jobs gains a fair amount since the cache hit happens between workflow runs, not just between jobs, and avoids needing an `install` job at all

File: ci/dhis2-verify-lib.yml (hunk 1 above)
Comment (varl) [reply ci/dhis2-verify-lib.yml:114]: Yep! Thanks. I was using cli-style as a test bed and that doesn't need a build step so that slipped out of my brain. Should we have a specific one for cli in addition to lib and app?

File: ci/dhis2-verify-app.yml (hunk 2 above)
Comment (varl) [reply ci/dhis2-verify-app.yml:17]: > > install job runs first, blocking parallel jobs (test, lint) until the cache has been created.
> 
> Were you able to test this? I'm not sure we gain much by saving and restoring the Yarn cache between jobs, as the install job itself still needs to run for each job and now it actually needs to run twice before any job can finish. Maybe we could cache `node_modules` in the `install` job and drop the `yarn install` repeated for each other job instead?

Yeah, I tested it and it works in some capacity, but in cli-style an install is so fast that it just saves 4 seconds (from 5 seconds to 1 second) so it is negligible. The "win" would really be in application workflow runs.

I like the idea of caching the node_modules, we can try that instead.

> Alternatively, just having the cache action and `yarn install` for each of the `lint`/`test`/`build`/`publish` jobs gains a fair amount since the cache hit happens between workflow runs, not just between jobs, and avoids needing an `install` job at all

Yeah, I thought about this too. The majority of our pull requests are Dependabot PRs, and those are going to be cache misses since they always (well ...) change the `yarn.lock` file.

So in those cases the parallel jobs would always run install `n` times, where `n` is the number of jobs in parallel. And each of the Dependabot PRs that rebases gets a new `yarn.lock` as well and triggers `n` runs, and so forth for all the open Dependabot PRs.

We can easily have 10:1 of Dependabot PRs vs. Human PRs, so I figure we should optimize for the 10x case if we want to save build minutes and :evergreen_tree:.

If you disagree with this rationale, we can go for omitting the "install" job completely.

File: ci/dhis2-verify-app.yml (hunk 3 above)
Comment (varl) [reply ci/dhis2-verify-app.yml:167]: Yep, you're right, good catch. I've done that for App Hub when passing around images so I'll see if it can be replicated here.

**Comment by varl** [comment]
> > Actions cannot be defined using workflow definitions in YAML, they need to be programmatic (JavaScript), which would be the next iteration to explore.
> 
> I _think_ we can use a [composite run steps action](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action) for this? (Haven't tested or looked into it more than noting this doc)

Cool! Going to take a look at this.

> > install job runs first, blocking parallel jobs (test, lint) until the cache has been created.
> 
> Were you able to test this? I'm not sure we gain much by saving and restoring the Yarn cache between jobs, as the install job itself still needs to run for each job and now it actually needs to run twice before any job can finish. Maybe we could cache `node_modules` in the `install` job and drop the `yarn install` repeated for each other job instead?

Yeah, I tested it and it works in some capacity, but in cli-style an install is so fast that it just saves 4 seconds (from 5 seconds to 1 second) so it is negligible. The "win" would really be in application workflow runs.

I like the idea of caching the node_modules, we can try that instead.

> Alternatively, just having the cache action and `yarn install` for each of the `lint`/`test`/`build`/`publish` jobs gains a fair amount since the cache hit happens between workflow runs, not just between jobs, and avoids needing an `install` job at all

Yeah, I thought about this too. The majority of our pull requests are Dependabot PRs, and those are going to be cache misses since they always (well ...) change the `yarn.lock` file.

So in those cases the parallel jobs would always run install `n` times, where `n` is the number of jobs in parallel. And each of the Dependabot PRs that rebases gets a new `yarn.lock` as well and triggers `n` runs, and so forth for all the open Dependabot PRs.

We can easily have 10:1 of Dependabot PRs vs. Human PRs, so I figure we should optimize for the 10x case if we want to save build minutes and :evergreen_tree:.

If you disagree with this rationale, we can go for omitting the "install" job completely.

**Comment by varl** [comment]
You can see the "cache node_modules" working here: https://github.com/dhis2/cli-style/actions/runs/306395532

It skips the install step completely if there is a cache hit. The entire workflow then runs in 51s, this is down from 1m 25s from the original version I made. :+1:

File: ci/dhis2-verify-app.yml [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+              if: steps.yarn-cache.outputs.cache-hit != 'true'
+              run: yarn install --frozen-lockfile
+
+            - name: Build
+              run: yarn build
+
+            - uses: actions/upload-artifact@v2
+              with:
+                  name: app-build
+                  path: '**/build'
```
Comment (varl) [inline ci/dhis2-verify-app.yml:75]: This is untested, but needed for monorepos.

File: ci/dhis2-verify-app.yml [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+              if: steps.yarn-cache.outputs.cache-hit != 'true'
+              run: yarn install --frozen-lockfile
+
+            - name: Test
+              run: yarn test
+
+    publish:
+        runs-on: ubuntu-latest
+        needs: [build, lint, test]
+        if: "!contains(github.event.head_commit.message, '[skip ci]')"
```
Comment (varl) [inline ci/dhis2-verify-app.yml:127]: Removed the `pull_request` case to allow us to deploy the feature branches to d2-ci.

File: ci/dhis2-verify-app.yml [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+              with:
+                  # uncomment following line for monorepo apps:
+                  #cwd: ./packages/app
+                  build-dir: build/app
+                  github-token: ${{ env.GH_TOKEN }}
+
+    release:
+        runs-on: ubuntu-latest
+        needs: [publish]
+        if: "github.event_name == 'push' && !contains(github.event.head_commit.message, '[skip ci]')"
```
Comment (varl) [inline ci/dhis2-verify-app.yml:153]: We do not want to create a release in GitHub for feature branches, only on the specified branches.

File: ci/dhis2-verify-lib.yml (hunk 1 above)
Comment (varl) [reply ci/dhis2-verify-lib.yml]: Made a new one, `dhis2-verify-node.yml`.

File: ci/dhis2-verify-app.yml [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+            - uses: actions/cache@v2
+              id: yarn-cache
+              with:
+                  path: '**/node_modules'
+                  key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+
+            - name: Install
+              if: steps.yarn-cache.outputs.cache-hit != 'true'
+              run: yarn install --frozen-lockfile
```
[the remark this answers is not in the document]
Comment (varl) [reply ci/dhis2-verify-app.yml:67]: I agree that it shouldn't be necessary. I had some problems related to this at first, but I think that after iterating a few times on these it should work without it. Let me try and remove it and see what happens in e.g. UI.

[the remark this answers is not in the document]
Comment (varl) [reply ci/dhis2-verify-app.yml:67]: Seems like it works now that we restore the entire node_modules and install the cypress binary explicitly:

https://github.com/dhis2/ui/runs/1273953364

[the remark this answers is not in the document]
Comment (varl) [reply ci/dhis2-verify-app.yml]: 972c1a1

[the remark this answers is not in the document]
Comment (amcgee) [reply ci/dhis2-verify-app.yml]: Nice!  So then we should remove the other `yarn install` steps?

For Cypress we should probably use the [official Cypress action](https://github.com/cypress-io/github-action) (which installs the binary) and should also definitely be running a matrix of Cypress jobs in parallel so that it doesn't take 12+ minutes (that run is still going), but that's probably outside the scope of this change (we should have a separate e2e workflow maybe...?  Or add it to the same one later?

[the remark this answers is not in the document]
Comment (amcgee) [reply ci/dhis2-verify-app.yml]: (I see you did this already :-) )

**Comment by varl** [comment]
> we should have a separate e2e workflow maybe...? Or add it to the same one later?

Undecided. I was thinking that we could add it to the same one, but comment it out by default.

**Comment by varl** [comment]
> > Actions cannot be defined using workflow definitions in YAML, they need to be programmatic (JavaScript), which would be the next iteration to explore.
> 
> I _think_ we can use a [composite run steps action](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action) for this? (Haven't tested or looked into it more than noting this doc)

Composite actions are quite a bit more limited than I would have liked (right now at least), so I could not get them working for us. The primary lacking feature is that a composite action cannot use `uses`. Only these props are available: https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/metadata-syntax-for-github-actions#runs-for-composite-run-steps-actions.

This basically limits us to:

```
runs:
  using: "composite"
  steps: 
    - run: yarn install
      shell: bash
  steps: 
    - run: yarn build
      shell: bash
  steps: 
    - run: yarn test
      shell: bash
```

But all the `uses` actions like checkout, cache, cypress, etc. would have to be done on the outside of the composite action.

Furthermore, composite actions don't support flow control such as `if`, so we could not decide in the composite action to _not_ run `install` if we have a cached copy.

I think we need to pass on `composite` actions it for now for this use case.

**Review by HendrikThePendric — COMMENTED** [review]
Great stuff! Left one comment that might need to be addressed.

File: ci/dhis2-verify-app.yml [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            # match branches in:
+            # https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#branches
+            - master
+            - next
+            - next-major
+            - alpha
+            - beta
+            - '[0-9]+.x'
+            - '[0-9]+.x.x'
+            - '[0-9]+.[0-9]+.x'
```
[the remark this answers is not in the document]
Comment (amcgee) [reply ci/dhis2-verify-app.yml:15]: @HendrikThePendric I think you're right... The [`npm` publish step](https://github.com/dhis2/workflows/pull/5/files/972c1a16602195cfdbeea7d065d5bab6ff9c242f#diff-47152e31368a6b5972e93ca5598ac3e5f3757a4ef0af210b28bc2a4945bd2ed5R100) uses Semantic Release, so only [semantic release's default release branches] are currently supported and we should be able to safely run it on other branches, but we also need to deploy to `d2-ci` on our app production branches...

File: ci/dhis2-verify-app.yml (hunk 4 above)
Comment (varl) [reply ci/dhis2-verify-app.yml]: Finally solved this.

File: ci/dhis2-verify-app.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (varl) [reply ci/dhis2-verify-app.yml:15]: e517cbd

**Comment by varl** [comment]
OK, this should be pretty much complete now. Tested in cli-style (node), ui (lib, monorepo and build), and usage analytics (app, build).

Untested is how this works for e.g. data-visualizer which is an an app with a monorepo, that deploys an app to d2-ci and publishes a plugin to npm.

**Comment by varl** [comment]
Added commented out e2e job config for apps and libs.

**Review by amcgee — APPROVED** [review]
LGTM!
